okhttp-testing-support/src/main/java/okhttp3/OkHttpClientTestRule.kt

```diff
@@ -55,25 +55,51 @@ class OkHttpClientTestRule : TestRule {
         acquireClient()
         try {
           base.evaluate()
+          logEventsIfFlaky()
+        } catch (t: Throwable) {
+          logEventsOnFailure(t)
+          throw t
         } finally {
           ensureAllConnectionsReleased()
           releaseClient()
         }
       }
 
       private fun acquireClient() {
-        prototype = prototypes.poll() ?: OkHttpClient.Builder()
-            .dns(SINGLE_INET_ADDRESS_DNS) // Prevent unexpected fallback addresses.
-            .build()
+        prototype = prototypes.poll() ?: freshClient()
+        clientEventsMap[prototype!!] = mutableListOf()
       }
 
       private fun releaseClient() {
         prototypes.push(prototype)
+        clientEventsMap.remove(prototype)
         prototype = null
       }
     }
   }
 
+  private fun logEventsIfFlaky() {
+    // TODO check if test is a known flaky test that should always log
+//    logEvents()
+  }
+
+  private fun logEventsOnFailure(t: Throwable) {
+    logEvents()
+  }
+
+  private fun logEvents() {
+    val events = clientEventsMap[prototype]
+
+    // Will fail if test overrides the listener
```

Comment:
This is pretty common, but once a test is recording events, we should avoid causing additional noise.